android/src/main/kotlin/com/shopify/reactnative/flash_list/AutoLayoutView.kt

```diff
@@ -66,6 +70,72 @@ class AutoLayoutView(context: Context) : ReactViewGroup(context) {
         }
     }
 
+    /** Fixes footer position along with rest of the items */
+    private fun fixFooter() {
+        if (!disableAutoLayout) {
+            val parentSV = getParentScrollView()
+            if (parentSV != null) {
+                val isAutoLayoutEndVisible = if (alShadow.horizontal) right <= parentSV.width else bottom <= parentSV.height
+                if (isAutoLayoutEndVisible) {
+                    val footer = getFooter();
+                    val diff = getFooterDiff()
+                    if (diff != 0 && footer != null) {
+                        val alParent = parent as View;
+                        if (alShadow.horizontal) {
+                            footer.offsetLeftAndRight(diff)
+                            right += diff
+                            alParent?.right += diff
+                        } else {
+                            footer.offsetTopAndBottom(diff)
+                            bottom += diff
+                            alParent?.bottom += diff
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getFooterDiff(): Int {
+        if (childCount == 0) {
+            alShadow.lastMaxBoundOverall = 0
+        } else if (childCount == 1) {
+            val firstChild = getChildAt(0)
+            alShadow.lastMaxBoundOverall = if (alShadow.horizontal) {
+                firstChild.right
+            } else {
+                firstChild.bottom
+            }
+        }
+        val autoLayoutEnd = if (alShadow.horizontal) right - left else bottom - top
+        return alShadow.lastMaxBoundOverall - autoLayoutEnd
+    }
+
+    private fun getFooter(): View? {
+        return (parent as ViewGroup)?.let {
+            val count = it.childCount;
+            for (i in 0 until count) {
+                val view = it.getChildAt(i)
+                if (view is CellContainer && view.index == -1) {
+                    return@let view as View
+                }
+            }
+            return@let null
+        }
+    }
+
+    private fun getParentScrollView(): View? {
+        var alParent = parent;
+        while (alParent != null) {
+            if (alParent is ScrollView || alParent is HorizontalScrollView) {
+                return alParent as View
+            }
+            alParent = alParent.parent;
+        }
+        return null
```

Comment:
```suggestion
        return generateSequence(parent) { it.parent }
                .firstOrNull { it is ScrollView || it is HorizontalScrollView } as? View
```

You could also rewrite it with this less-verbose version which would be more similar to what we do on iOS. It'd be nice to keep those two implementations as close as possible.

Reply:
Good suggestion. I'm skipping for now just to understand these sequences better. Nothing beats while when it comes to speed.

Reply:
I dug a little deeper to see the Kotlin Bytecode, just out of interest:

This is bytecode for the version with `while`:
```
   L0
    LINENUMBER 129 L0
    ALOAD 0
    INVOKEVIRTUAL com/shopify/reactnative/flash_list/AutoLayoutView.getParent ()Landroid/view/ViewParent;
    ASTORE 1
   L1
    LINENUMBER 130 L1
   L2
    ALOAD 1
    IFNULL L3
   L4
    LINENUMBER 131 L4
    ALOAD 1
    INSTANCEOF android/widget/ScrollView
    IFNE L5
    ALOAD 1
    INSTANCEOF android/widget/HorizontalScrollView
    IFEQ L6
   L5
    LINENUMBER 132 L5
    ALOAD 1
   L7
    CHECKCAST android/view/View
    ARETURN
   L6
    LINENUMBER 134 L6
    ALOAD 1
    INVOKEINTERFACE android/view/ViewParent.getParent ()Landroid/view/ViewParent; (itf)
    ASTORE 1
   L8
    LINENUMBER 130 L8
    GOTO L2
   L3
    LINENUMBER 136 L3
    ACONST_NULL
    ARETURN
   L9
```

This is bytecode with the sequence:

```
   L0
    LINENUMBER 129 L0
   L1
    LINENUMBER 130 L1
   L2
    LINENUMBER 129 L2
    ALOAD 0
    INVOKEVIRTUAL com/shopify/reactnative/flash_list/AutoLayoutView.getParent ()Landroid/view/ViewParent;
    GETSTATIC com/shopify/reactnative/flash_list/AutoLayoutView$getParentScrollView$1.INSTANCE : Lcom/shopify/reactnative/flash_list/AutoLayoutView$getParentScrollView$1;
    CHECKCAST kotlin/jvm/functions/Function1
    INVOKESTATIC kotlin/sequences/SequencesKt.generateSequence (Ljava/lang/Object;Lkotlin/jvm/functions/Function1;)Lkotlin/sequences/Sequence;
    ASTORE 1
   L3
    LINENUMBER 130 L3
    NOP
   L4
    ICONST_0
    ISTORE 2
   L5
```

Additionally, you can decompile the bytecode to Java.

For `while`, Java code looks the following:

```
for(ViewParent autoLayoutParent = this.getParent(); autoLayoutParent != null; autoLayoutParent = autoLayoutParent.getParent()) {  
   if (autoLayoutParent instanceof ScrollView || autoLayoutParent instanceof HorizontalScrollView) {  
      return (View)autoLayoutParent;  
   }  
}  
  
return null;
```

And for the sequence, it looks like this:
```
Sequence $this$firstOrNull$iv = SequencesKt.generateSequence(this.getParent(), (Function1)null.INSTANCE);  
int $i$f$firstOrNull = false;  
Iterator var3 = $this$firstOrNull$iv.iterator();  
  
Object var10000;  
while(true) {  
   if (!var3.hasNext()) {  
      var10000 = null;  
      break;  
   }  
  
   Object element$iv = var3.next();  
   ViewParent it = (ViewParent)element$iv;  
   int var6 = false;  
   if (it instanceof ScrollView || it instanceof HorizontalScrollView) {  
      var10000 = element$iv;  
      break;  
   }  
}  
  
if (!(var10000 instanceof View)) {  
   var10000 = null;  
}  
  
return (View)var10000;
```

I haven't dug deeper to see the Java bytecode but it's fair to say the sequence will be slower. However, since it compiles to `while`, the results will be most likely similar and it's more readable. I'd prefer `generateSequence` in case the readability win outweighs the small performance hit but in this case, even the `while` alternative is pretty easy to read, so we can keep it as-is 👍